CARDS-1637: Make certain answers easily accessible in the subject json

veronikaslc opened the pull request.

Commit:
CARDS-1637: Make certain answers easily accessible in the subject json

veronikaslc committed:
CARDS-1637: Make certain answers easily accessible in the subject json

Minor code style change

Co-authored-by: Marta Girdea <marta@phenotips.org>

Commit:
CARDS-1637: Make certain answers easily accessible in the subject json
Addressed codereview comment: small change in config file structure

sdumitriu committed:
CARDS-1637: Make certain answers easily accessible in the subject json
Bugfix: subject copying never happened because the wrong subject type name is used
Bugfix: copying answers from a parent subject form doesn't work
Refactoring: it is possible to have dependencies injected in a base abstract class; do that to avoid lots of code duplication
Refactoring: configuration path should be `/apps/cards/config/CopyAnswers/SubjectTypes/Patient/Visit` instead of `/apps/cards/config/CopyAnswers/Visit/Visit`

sdumitriu committed:
CARDS-1637: Make certain answers easily accessible in the subject json
Bugfix: sometimes some answers are copied to the subject's parent
Refactoring: move more code up to the abstract class

sdumitriu committed:
[misc] Simplify the code a bit and add more explanatory comments

sdumitriu committed:
[misc] Reduce method visibility

sdumitriu committed:
[misc] Better method name

sdumitriu committed:
[misc] Javadoc improvements

sdumitriu committed:
[misc] Move services to the class where they are used

sashaandjic commented:
I do not see all the values from my visit. Surveys, Provider, Location and Status is missing.
![image](https://user-images.githubusercontent.com/23243244/161143192-c41b3873-3316-4baf-ad4c-e516a30e3de7.png)
![image](https://user-images.githubusercontent.com/23243244/161143231-8f543069-0ebd-460a-b7e6-3c74d3682b0f.png)

marta- commented:
> I do not see all the values from my visit. Values for Surveys, Provider, Location and Status are missing.

That's expected, only some fields are configured to be copied:
```
{
  "jcr:primaryType": "nt:unstructured",
  "jcr:reference:time": "/Questionnaires/Visit information/time",
  "jcr:reference:surveys_complete": "/Questionnaires/Visit information/surveys_complete",
  "jcr:reference:surveys_submitted": "/Questionnaires/Visit information/surveys_submitted",
  "jcr:reference:mrn": "/Questionnaires/Patient information/mrn",
  "jcr:reference:first_name": "/Questionnaires/Patient information/first_name",
  "jcr:reference:last_name": "/Questionnaires/Patient information/last_name"
}
```
Looks like what you get matches the list of copied fields, so all's good.

marta- commented:
@sashaandjic One more thing to test: this PR involves some refactoring that affects the PROMs dashboard; please do a quick dashboard test to ensure visits and surveys are displayed as expected.

sashaandjic commented:
Deleted some visits, and created new ones (past, present, future). Now all of them are showing up in the `Past` and in the `Today` tab

https://user-images.githubusercontent.com/23243244/161156373-986c3263-597c-4d82-86ee-168db0cd3a0b.mp4

.

Commit:
CARDS-1637: Make certain answers easily accessible in the subject json
Removed unused import

veronikaslc commented:
I have added one ommeit for removing unised import and rebased on the latest dev. I see that hte last comments from Sasha are actually an expected behaviour, so I had removed the `bug` label and added `Test me` one.

sashaandjic commented:
I created 2 visits 2 weeks apart and got all 5 surveys generated for both - I expected none for the second one. 
Another visit 2 moth apart got 5 surveys also generated - I expected only 3.

marta- commented:
> Hope this is out of scope and not introduced in this PR. I created 2 visits 2 weeks apart and got all 5 surveys generated for both - I expected none for the second one. Another visit 2 moth apart got 5 surveys also generated - I expected only 3.

Sounds completely unrelated, this PR does not affect the survey frequency logic. Please try to replicate on `dev`.

sashaandjic commented:
> Sounds completely unrelated, this PR does not affect the survey frequency logic. Please try to replicate on `dev`.

Replicated on dev and logged as regression.

sashaandjic reviewed (approved):
tested
